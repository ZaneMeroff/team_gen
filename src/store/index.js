import Vue from "vue"
import Vuex from "vuex"

Vue.use(Vuex)

export default new Vuex.Store({
  state: {
    teamNum: 2,
    playerNum: 2,
    playerList: ["", ""],
  },

  mutations: {

    plusteamNum: (state) => state.teamNum++,
    minusteamNum: (state) => state.teamNum !== 2 && state.teamNum--,

    plusplayerNum: (state) => state.playerNum++,
    minusplayerNum: (state) => state.playerNum !== 2 && state.playerNum--,

    updatePlayerList: (state, data) => state.playerList = data,
    updatePlayerName: (state, data) => state.playerList[data.id] = data.name,
  },

  actions: {},
  modules: {},
})
