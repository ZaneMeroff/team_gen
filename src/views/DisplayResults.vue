<template>
  <div>

    <router-link class="router-link" to="/enterNames">
      <ArrowBtn :arrowDirection="'left'"/>
    </router-link>

    <ArrowBtn
      :disabled="true"
      :arrowDirection="'right'"
    />

    <TeamCard
      v-for="(card, index) in results"
      :key="index"
      :members="card.members"
      :teamId="card.id"
    />

  </div>
</template>

<script>
  // need spec with snapshot

  import ArrowBtn from "../components/ArrowBtn"
  import TeamCard from "../components/TeamCard"

  export default {
    name: "DisplayResults",
    components: {
      ArrowBtn,
      TeamCard,
    },
    data() {
      return {
        shuffledPlayerList: [],
        emptyTeams: [],
        playersPerTeam: 0,
        stepCounter: 0,
        results: [],
      }
    },
    methods: {

      shuffle(array) {
        let currentIndex = array.length, temporaryValue, randomIndex
        while (0 !== currentIndex) {
          randomIndex = Math.floor(Math.random() * currentIndex)
          currentIndex--
          temporaryValue = array[currentIndex]
          array[currentIndex] = array[randomIndex]
          array[randomIndex] = temporaryValue
        }
        this.shuffledPlayerList = array
        this.stepCounter++
      },

      buildTeamObjs(num) {
        let teams = []
        for ( let i = 1; i <= num; i++ ) {
          teams.push({ id: i, members: [] })
        }
        this.emptyTeams = teams
        this.stepCounter++
      },

      getPlayersPerTeam(names, teamNum) {
        this.playersPerTeam = Math.floor(names.length / teamNum)
        this.stepCounter++
      },

      clone(data) {
        return JSON.parse(JSON.stringify(data))
      },

      assignTeams(names, teams) {
        teams.forEach(team => {
          for ( let i = 1; i <= this.playersPerTeam; i++ ) {
            team.members.push(names.pop())
          }
        })
        teams.forEach(team => names.length && team.members.push(names.pop()))
        this.results = teams
      },
    },
    computed: {

      playerList() {
        return this.$store.state.playerList
      },
      teamNum() {
        return this.$store.state.teamNum
      },
    },
    watch: {
      stepCounter() {
        this.stepCounter === 3 && this.assignTeams(this.shuffledPlayerList, this.emptyTeams)
      }
    },
    mounted() {
      this.shuffle(this.clone(this.playerList))
      this.buildTeamObjs(this.teamNum)
      this.getPlayersPerTeam(this.playerList, this.teamNum)
    },
  }
</script>

<style scoped>

</style>
