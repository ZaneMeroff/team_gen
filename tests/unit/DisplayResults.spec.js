import { createLocalVue, shallowMount } from '@vue/test-utils'
import store from "@/store/index"
import Vuex from "vuex"
import VueRouter from "vue-router"
import DisplayResults from "@/views/DisplayResults"

// --------- Vuex & Router setup ------------

const localVue = createLocalVue()
localVue.use(Vuex)
localVue.use(VueRouter)

// --------------- mocks --------------------

jest.spyOn(global.Math, "random").mockReturnValue(.5)

// ------------------------------------------

describe("DisplayResults", () => {

  describe("snapshots", () => {

    it("should render with default html", () => {
      const component = shallowMount(DisplayResults, { store, localVue })

      expect(component.element).toMatchSnapshot()
    })
  })

  describe("data", () => {

    describe("shuffledPlayerList", () => {

      it("should return a default value", () => {
        const component = shallowMount(DisplayResults, { store, localVue })

        expect(component.vm.$data.shuffledPlayerList).toEqual(["",""])
      })
    })

    describe("emptyTeams", () => {

      it("should return a default value", () => {
        const component = shallowMount(DisplayResults, { store, localVue })

        const expected = [
          {id: 1, members: [] },
          {id: 2, members: [] },
        ]
        expect(component.vm.$data.emptyTeams).toEqual(expected)
      })
    })

    describe("playersPerTeam", () => {

      it("should return a default value", () => {
        const component = shallowMount(DisplayResults, { store, localVue })

        expect(component.vm.$data.playersPerTeam).toEqual(1)
      })
    })

    describe("stepCounter", () => {

      it("should return a default value", () => {
        const component = shallowMount(DisplayResults, { store, localVue })

        expect(component.vm.$data.stepCounter).toEqual(3)
      })
    })

    describe("results", () => {

      it("should return a default value", () => {
        const component = shallowMount(DisplayResults, { store, localVue })

        expect(component.vm.$data.results).toEqual([])
      })
    })
  })

  describe("methods", () => {

    describe("shuffle", () => {

      it("should shuffle array", () => {
        const component = shallowMount(DisplayResults, { store, localVue })

        component.vm.shuffle([ "A", "B", "C", "D" ])

        const expected = [ "A", "D", "B", "C" ]
        expect(component.vm.$data.shuffledPlayerList).toEqual(expected)
      })
    })

    describe("buildTeamObjs", () => {

      it("should build correct number of team objs", () => {
        const component = shallowMount(DisplayResults, { store, localVue })

        component.vm.buildTeamObjs(3)

        const expected = [
          { id: 1, members: [] },
          { id: 2, members: [] },
          { id: 3, members: [] },
        ]
        expect(component.vm.$data.emptyTeams).toEqual(expected)
      })

      describe("getPlayersPerTeam", () => {

        it("should correctly set playesrPerTeam", () => {
          const component = shallowMount(DisplayResults, { store, localVue })

          const names = [ "Bob", "Sue", "Abe", "Dee", "Ben", "Dot"]
          component.vm.getPlayersPerTeam(names, 2)
          expect(component.vm.$data.playersPerTeam).toEqual(3)
        })

        it("should correctly set playesrPerTeam", () => {
          const component = shallowMount(DisplayResults, { store, localVue })

          const names = [ "Bob", "Sue", "Abe", "Dee", "Ben"]
          component.vm.getPlayersPerTeam(names, 2)
          expect(component.vm.$data.playersPerTeam).toEqual(2)
        })

        it("should correctly set playesrPerTeam", () => {
          const component = shallowMount(DisplayResults, { store, localVue })

          const names = [ "Bob", "Sue", "Abe" ]
          component.vm.getPlayersPerTeam(names, 3)
          expect(component.vm.$data.playersPerTeam).toEqual(1)
        })
      })

      describe("clone", () => {

        it("should return clone of data", () => {
          const component = shallowMount(DisplayResults, { store, localVue })
          const names = [ "Bob", "Sue", "Abe" ]

          const result = component.vm.clone(names)

          expect(result).toEqual(names)
        })
      })

      describe("assignTeams", () => {

        it("should assign teams evenly", () => {
          const component = shallowMount(DisplayResults, { store, localVue })
          const names = [ "A", "B", "C", "D", "E", "F" ]
          const teams = [
            { id: 1, members: [] },
            { id: 2, members: [] },
            { id: 3, members: [] },
          ]

          component.vm.assignTeams(names, teams)

          const expected = [
            { id: 1, members: [ "F", "C" ] },
            { id: 2, members: [ "E", "B" ] },
            { id: 3, members: [ "D", "A" ] },
          ]
          expect(component.vm.$data.results).toEqual(expected)
        })
      })
    })
  })

  describe("computed properties", () => {

    describe("playerList", () => {

      it("should return default playerList", () => {
        const component = shallowMount(DisplayResults, { store, localVue })

        const expected = ["", ""]
        expect(component.vm.playerList).toEqual(expected)
      })
    })

    describe("teamNum", () => {

      it("should return default teamNum", () => {
        const component = shallowMount(DisplayResults, { store, localVue })

        const expected = 2
        expect(component.vm.teamNum).toEqual(expected)
      })
    })
  })

  describe("watch", () => {

    describe("stepCounter", () => {

      it("should call assignTeams when stepCounter is 3", async () => {
        const component = shallowMount(DisplayResults, { store, localVue })
        const spy = jest.spyOn(component.vm, "assignTeams")

        await expect(component.vm.$data.stepCounter).toEqual(3)
        expect(spy).toHaveBeenCalledTimes(1)
      })
    })
  })

  describe("lifecycle hooks", () => {

    describe("mounted", () => {

      it("should call shuffle", () => {
        const spy = jest.spyOn(DisplayResults.methods, "shuffle")
        const component = shallowMount(DisplayResults, { store, localVue })

        expect(spy).toHaveBeenCalledTimes(1)
      })

      it("should call buildTeamObjs", () => {
        const spy = jest.spyOn(DisplayResults.methods, "buildTeamObjs")
        const component = shallowMount(DisplayResults, { store, localVue })

        expect(spy).toHaveBeenCalledTimes(1)
      })

      it("should call getPlayersPerTeam", () => {
        const spy = jest.spyOn(DisplayResults.methods, "getPlayersPerTeam")
        const component = shallowMount(DisplayResults, { store, localVue })

        expect(spy).toHaveBeenCalledTimes(1)
      })
    })
  })
})
