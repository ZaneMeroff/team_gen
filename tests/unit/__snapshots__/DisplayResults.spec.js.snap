// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DisplayResults snapshots should render with default html 1`] = `
<div
  class="display-results-container"
>
  <div
    class="nav-btn-container"
  >
    <router-link-stub
      ariacurrentvalue="page"
      class="router-link"
      event="click"
      tag="a"
      to="/enterNames"
    >
      <arrowbtn-stub
        arrowdirection="left"
      />
    </router-link-stub>
     
    <arrowbtn-stub
      arrowdirection="right"
      disabled="true"
    />
  </div>
   
  <div
    class="team-cards-container"
  />
</div>
`;
